iwyu.cc

```diff
@@ -3478,21 +3510,17 @@ class IwyuAstConsumer
     // information here.  Note: we have to store this even if this is
     // an ast node we would otherwise ignore, since other AST nodes
     // (which we might not ignore) can depend on it.
-    for (UsingDecl::shadow_iterator it = decl->shadow_begin();
-         it != decl->shadow_end(); ++it) {
-      AddUsingDeclaration((*it)->getTargetDecl(), decl);
-    }
-
-    if (CanIgnoreCurrentASTNode())  return true;
+    AddShadowDeclarations(decl);
 
     // The shadow decls hold the declarations for the var/fn/etc we're
     // using.  (There may be more than one if, say, we're using an
-    // overloaded function.)  We check to make sure nothing we're
-    // using is an iwyu violation.
-    for (UsingDecl::shadow_iterator it = decl->shadow_begin();
-         it != decl->shadow_end(); ++it) {
-      ReportDeclForwardDeclareUse(CurrentLoc(), (*it)->getTargetDecl());
-    }
+    // overloaded function.)  We don't want to add all of them at once
+    // though, because that will drag in every overload even if we're
+    // only using one.  Instead, we keep track of the using decl and
+    // mark it as touched when something actually uses it.
+    preprocessor_info().FileInfoFor(CurrentFileEntry())->AddUsingDecl(decl);
+
+    if (CanIgnoreCurrentASTNode())  return true;
```

Comment:
What was the reason to move `if (CanIgnoreCurrentASTNode())  return true;` from before `ReportDeclForwardDeclareUse` after `AddUsingDecl`? Because when I move it back, nothing fails.